packages/block-library/src/navigation/edit.js

```diff
@@ -366,6 +371,12 @@ export default compose( [
 			selectedBlockId,
 		] )?.length;
 
+		const onlyLinkInnerBlocks = innerBlocks.every(
+			( block ) =>
+				block.name === 'core/navigation-link' ||
+				block.name === 'core/navigation-submenu'
+		);
```

Comment:
Have we considered allowing `__experimentalDirectInsert` to be a match function which returns true or false?  It doesn't look like we need any other information specific to the render function. One benefit is that we can then make it a constant and can define it next to the default block.

For example it might look like:

```jsx
const DIRECT_INSERT = ( block ) => { 
    return block.innerBlocks.every( ( { name } ) => name === 'core/navigation-link' || name === 'core/navigation-submenu' );
};
```

And in the selector we can update it to call the match function.

Reply:
Oooh nice! I'll try it out.

Reply:
Hmm, this doesn't work well when we either add or remove other block types, because the selector won't re-run with every block update, so it won't know when the content has changed. Whereas if we keep the logic inside the edit function it updates with every change in the inner blocks.

Reply:
Have you tried without `createSelector`? We should get proper results provided we recalculate whenever the block.innerBlock reference updates.

Reply:
The following works for me. (I didn't notice any infinite loops, but let me know if there's a specific case to test.)

Non-memo'd:

```jsx
export const __experimentalGetDirectInsertBlock = (
	state,
	rootClientId = null
) => {
	if ( ! rootClientId ) {
		return;
	}
	const settings = state.blockListSettings[ rootClientId ];
	const defaultBlock = settings?.__experimentalDefaultBlock;
	if ( ! defaultBlock ) {
		return;
	}
	const directInsert = settings?.__experimentalDirectInsert(
		getBlock( state, rootClientId )
	);
	if ( ! directInsert ) {
		return;
	}

	return defaultBlock;
};
```

Or memo'd

```jsx
export const __experimentalGetDirectInsertBlock = createSelector(
	( state, rootClientId = null ) => {
		if ( ! rootClientId ) {
			return;
		}
		const settings = state.blockListSettings[ rootClientId ];
		const defaultBlock = settings?.__experimentalDefaultBlock;
		if ( ! defaultBlock ) {
			return;
		}
		const directInsert = settings?.__experimentalDirectInsert(
			getBlock( state, rootClientId )
		);
		if ( ! directInsert ) {
			return;
		}

		return defaultBlock;
	},
	( state, rootClientId ) => [
		state.blockListSettings[ rootClientId ],
		state.blocks.tree[ rootClientId ],
	]
);
```

Reply:
Oh I see, I was forgetting to add `state.blocks.tree[ rootClientId ]` to the dependencies 🤦 

I think we should also accept a boolean value though, to avoid having to do something like `() => true` when we want to _always_ insert the default block.

Reply:
> I think we should also accept a boolean value though

That sounds reasonable to me 👍